### ISSUE TITLE
[require-await] require-await false positive 

### ISSUE THREAD

**Issue by mdouglass** [typescript-eslint/typescript-eslint#1188]
**Repro**

```JSON
{
  "parser": "@typescript-eslint/parser",
  "parserOptions": { "project": "./tsconfig.json" },
  "plugins": ["@typescript-eslint"],
  "rules": {
    "@typescript-eslint/require-await": "error"
  }
}
```

```TS
async function testFunction(): Promise<void> {
  await Promise.all([1, 2, 3].map(async value => Promise.resolve(value)))
}
```

**Expected Result**
No warnings/errors

**Actual Result**
  1:1  error  Async function 'testFunction' has no 'await' expression  @typescript-eslint/require-await

**Additional Info**
Changing the function passed to map to `value => value` causes the false positive to not trigger.

**Versions**

| package | version |
| ---------------------------------- | ------- |
| @typescript-eslint/eslint-plugin | 2.6.1 |
| @typescript-eslint/parser | 2.6.1 |
| TypeScript | 3.7.2 |
| ESLint | 6.6.0 |
| node | 13.1.0 |
| npm | 6.12.1 |

**Comment by bradzacher** [comment]
It's a bug in the rule implementation.
The extension knows the inner callback doesn't need have an error, so it doesn't call the base rule's handler.
But in doing so, it means the base rule doesn't properly update its internal state, so it ends up looking at the state for the cb whilst exiting `testFunction`.
